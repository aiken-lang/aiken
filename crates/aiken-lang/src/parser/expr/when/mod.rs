use chumsky::prelude::*;

mod clause;
mod guard;

pub use clause::parser as clause;
pub use guard::parser as guard;

use crate::{
    expr::UntypedExpr,
    parser::{error::ParseError, token::Token},
};

pub fn parser(
    expression: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, UntypedExpr, Error = ParseError> + '_ {
    just(Token::When)
        // TODO: If subject is empty we should return ParseErrorType::ExpectedExpr,
        .ignore_then(expression.clone().map(Box::new))
        .then_ignore(just(Token::Is))
        .then_ignore(just(Token::LeftBrace))
        // TODO: If clauses are empty we should return ParseErrorType::NoCaseClause
        .then(clause(expression).repeated())
        .then_ignore(just(Token::RightBrace))
        .map_with_span(|(subject, clauses), span| UntypedExpr::When {
            location: span,
            subject,
            clauses,
        })
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn when_basic() {
        assert_expr!(
            r#"
            when a is {
              2 if x > 1 -> 3
              1 | 4 | 5 -> {
                let amazing = 5
                amazing
              }
              3 -> 9
              _ -> 4
            }
            "#
        );
    }
}
