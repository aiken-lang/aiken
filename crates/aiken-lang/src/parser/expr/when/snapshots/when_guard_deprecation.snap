---
source: crates/aiken-lang/src/parser/expr/when/mod.rs
description: "Invalid code (parse error):\n\nwhen a is {\n  2 if x > 1 -> 3\n  _ -> 1\n}\n"
---
[
    ParseError {
        kind: DeprecatedWhenClause,
        span: 14..29,
        while_parsing: None,
        expected: {},
        label: Some(
            "deprecated",
        ),
    },
]
